app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -44,38 +49,54 @@ def callback_for(provider)
       cta_variant: cta_variant,
     )
 
-    if persisted_and_valid?
+    if user_persisted_and_valid?
+      # TODO: [thepracticaldev/oss] remove this
       # delete legacy session based cookie once the user has logged in again
       request.cookie_jar.delete(LEGACY_COOKIE_NAME) if request.cookies[LEGACY_COOKIE_NAME]
 
       remember_me(@user)
-      sign_in_and_redirect @user, event: :authentication
+
       set_flash_message(:notice, :success, kind: provider.to_s.capitalize) if is_navigational_format?
-    elsif persisted_but_username_taken?
+
+      sign_in_and_redirect(@user, event: :authentication)
+    # NOTE: I can't find a way to test this path
+    # as `User` will assign a temporary username if the username already exists
+    # see https://github.com/thepracticaldev/dev.to/blob/27131f6f420df347a467f8e9afc84a6af2fcb13a/app/models/user.rb#L532-L555
+    elsif user_persisted_but_username_taken?
       redirect_to "/settings?state=previous-registration"
+    # NOTE: I can't find a way to test this path
+    # as `Authentication::Authenticator.call` invokes `User.save!` which will
+    # raise errors for a validation error.
+    # In the past we had 1 path (update_user) which would have ended up
+    # here in case of validation errors, see:
+    # https://github.com/thepracticaldev/dev.to/blob/80737b540453afe8775128cb37bd379b7c09c7e8/app/services/authorization_service.rb#L77
     else
       session["devise.#{provider}_data"] = request.env["omniauth.auth"]
+
       user_errors = @user.errors.full_messages
-      Rails.logger.error "Log in error: sign in failed. username: #{@user.username} - email: #{@user.email}"
-      Rails.logger.error "Log in error: auth data hash - #{request.env['omniauth.auth']}"
-      Rails.logger.error "Log in error: auth data hash - #{request.env['omniauth.error']&.inspect}"
-      Rails.logger.error "Log in error: user_errors: #{user_errors}"
+
+      Rails.logger.error("Log in error: sign in failed. username: #{@user.username} - email: #{@user.email}")
+      Rails.logger.error("Log in error: auth data hash - #{request.env['omniauth.auth']}")
+      Rails.logger.error("Log in error: auth data hash - #{request.env['omniauth.error']&.inspect}")
+      Rails.logger.error("Log in error: user_errors: #{user_errors}")
+
       flash[:alert] = user_errors
       redirect_to new_user_registration_url
     end
   rescue StandardError => e
-    Rails.logger.error "Log in error: #{e}"
-    Rails.logger.error "Log in error: auth data hash - #{request.env['omniauth.auth']}"
-    Rails.logger.error "Log in error: auth data hash - #{request.env['omniauth.error']&.inspect}"
+    Rails.logger.error("Log in error: #{e}")
+    Rails.logger.error("Log in error: auth data hash - #{request.env['omniauth.auth']}")
+    Rails.logger.error("Log in error: auth data hash - #{request.env['omniauth.error']&.inspect}")
```

Comment:
This is more for my own edification, but why do we log to the Rails logger here instead of sending this data to DataDog? (Maybe this is a silly question but seems like when we rescue errors we tend to send them to DataDog?)

Reply:
Asked myself the same question, @mstruve should we change that? I would but maybe not in this PR. We also need to figure out if we actually use those two logic paths we don't seem to use at all

Reply:
The next project on my list for SRE is to move our logging from Timber to Datadog so everyone can see it and then come up with a logging framework that makes things more searchable beyond message strings. Lets leave this for now, I will be overhauling the logging soon. Anything we want to give access to the whole team immediately feel free to send to Datadog as a metric.